src/osmet/io_helpers.rs

```diff
@@ -0,0 +1,191 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::convert::{TryFrom, TryInto};
+use std::fs::OpenOptions;
+use std::io::{self, ErrorKind, Read, Write};
+use std::os::unix::io::AsRawFd;
+use std::path::Path;
+
+use error_chain::bail;
+use openssl::hash::{Hasher, MessageDigest};
+use serde::{Deserialize, Serialize};
+
+use super::*;
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Clone, Default)]
+pub struct Sha256Digest([u8; 32]);
+
+impl TryFrom<Hasher> for Sha256Digest {
+    type Error = Error;
+
+    fn try_from(mut hasher: Hasher) -> std::result::Result<Self, Self::Error> {
+        let digest = hasher.finish().chain_err(|| "finishing hash")?;
+        Ok(Sha256Digest(
+            digest
+                .as_ref()
+                .try_into()
+                .chain_err(|| "converting to SHA256")?,
+        ))
+    }
+}
+
+/// This is like `std::io:copy()`, but limits the number of characters copied over. The `Read`
+/// trait has `take()`, but that takes ownership of the reader. We also take a buf to avoid
+/// re-initializing a block each time (std::io::copy() gets around this by using MaybeUninit, but
+/// that requires using nightly and unsafe functions).
+pub fn copy_n(
+    reader: &mut impl Read,
+    writer: &mut impl Write,
+    mut n: u64,
+    buf: &mut [u8],
+) -> Result<u64> {
+    let mut written = 0;
+    loop {
+        if n == 0 {
+            return Ok(written);
+        }
+        let bufn = if n < (buf.len() as u64) {
+            &mut buf[..n as usize]
+        } else {
+            &mut buf[..]
+        };
+        let len = match reader.read(bufn) {
+            Ok(0) => return Ok(written),
+            Ok(len) => len,
+            Err(ref e) if e.kind() == ErrorKind::Interrupted => continue,
+            Err(e) => return Err(e.into()),
+        };
+        assert!(len as u64 <= n);
+        writer.write_all(&bufn[..len])?;
+        written += len as u64;
+        n -= len as u64;
+    }
+}
+
+/// This is like `copy_n()` but errors if the number of bytes copied is less than expected.
+pub fn copy_exactly_n(
+    reader: &mut impl Read,
+    writer: &mut impl Write,
+    n: u64,
+    buf: &mut [u8],
+) -> Result<u64> {
+    let bytes_copied = copy_n(reader, writer, n, buf)?;
+    if bytes_copied != n {
+        bail!("expected to copy {} but instead copied {}", n, bytes_copied);
+    }
+    Ok(n)
+}
+
+// ab/cdef....file --> 0xabcdef...
+pub fn object_path_to_checksum(path: &Path) -> Result<Sha256Digest> {
+    let chksum2 = path
+        .parent()
+        .unwrap()
+        .file_name()
+        .unwrap()
+        .to_str()
+        .unwrap();
+    assert_eq!(chksum2.len(), 2);
+    let chksum62 = path.file_stem().unwrap().to_str().unwrap();
+    assert_eq!(chksum62.len(), 62);
```

Comment:
Yeah, makes sense. (I mean, if the path isn't as expected, then there's an odd-looking file in the OSTree repo object store, which is super dubious. But it's silly to just throw that out instead of printing something out that will help us debug.)